functional_programming_tests/test/main.cpp

```diff
@@ -0,0 +1,250 @@
+    if (path.size() > 0) {
+      for (auto const& position : path) {
+        message.data.push_back(position.x);
+        message.data.push_back(position.y);
+      }
+    }
+    return message;
+  }
+
+};
+
+Map<unsigned char> set_costmap_callback(const std::shared_ptr<example_srvs::srv::SetMap::Request> request,
+                                        std::shared_ptr<example_srvs::srv::SetMap::Response> response) {
+  response->success.data = false;
+  Map<unsigned char> costmap;
+  const auto parsed_map = PathGeneratorUtilities::parseSetMapRequest(request);
+  if (parsed_map.has_value()) {
+    response->success.data = true;
+    costmap = parsed_map.value();
+  }
+  return costmap;
+}
+
+// TODO: The return of this function can be the response
+// TODO: Edit this function such that the generate_path function is an input
+// TODO: Set the return  to be a std::expected with the various errors ()
+// TODO: Make a function that takes the error
+std::shared_ptr<example_srvs::srv::GetPath::Response> generate_path_callback(
+                           const std::shared_ptr<example_srvs::srv::GetPath::Request> request,
+                           const Map<unsigned char>& costmap,
+                           std::function<std::optional<Path>(Position const&, Position const&, Map<unsigned char> const&)> path_generater_function)
+{
+  std::shared_ptr<example_srvs::srv::GetPath::Response> response = std::make_shared<example_srvs::srv::GetPath::Response>();
+  if (costmap.get_data().size() == 0) {
+    response->success.data = false;
+    response->path = std_msgs::msg::UInt8MultiArray();
+    return response;
+  }
+  // Check to make sure start and goal fields of the request are of size 2
+  if (request->start.data.size() != 2) {
+    response->success.data = false;
+    response->path = std_msgs::msg::UInt8MultiArray();
+    return response;
+  }
+  if (request->goal.data.size() != 2) {
+    response->success.data = false;
+    response->path = std_msgs::msg::UInt8MultiArray();
+    return response;
+  }
+
+  auto const start = Position{request->start.data[0], request->start.data[1]};
+  auto const goal = Position{request->goal.data[0], request->goal.data[1]};
+
+  // Generate the path using the path generator function that was input
+  auto const path = path_generater_function(start, goal, costmap);
+
+  // Start populating the response message
+  auto response_path = std_msgs::msg::UInt8MultiArray();
+
+  if (path.has_value()) {
+    response_path = PathGeneratorUtilities::createUInt8MultiArrayMessageFromPath(path.value());
+  }
+
+  response->success.data = path.has_value();
+  response->path = response_path;
+
+  return response;
+}
+
+struct MainObject {
+
+  MainObject() : set_costmap_wrapper_{
+    [this](const std::shared_ptr<example_srvs::srv::SetMap::Request> request,
+               std::shared_ptr<example_srvs::srv::SetMap::Response> response) -> void
+    {
+      this->map_ = set_costmap_callback(request, response);
```

Comment:
This would make a lot more sense if the entire body of `set_costmap_callback` was just in this lamba

Reply:
Moved the logic to this function.